index.js

```
@@ -1,301 +1,318 @@
-            var suffix = stats.birthtime.toISOString() + rand + that.suffix;
-
-            return that._rotate( suffix, function ( err ) {
-                if ( err ) {
-                    return cb( err );
-                }
-
-                that._reopen( function () {
-                    that._write.call( that, data, encoding, cb );
-                })
+/**
+ * @private
+ */
+function _upload(path, stream) {
+  return new Promise((resolve, reject) => {
+    if (!stream.toUpload) {
+      resolve(path)
+    } else {
+      let date = path.match(
+        /(\d{4})-(\d{2})-(\d{2})T\d{2}:\d{2}:\d{2}\.\d{3}Z/)
+      let fileName = pathlib.basename(path)
+      let s3 = stream._s3
+      let key = pathlib.join(
+        s3.prefix,
+        date[1],
+        date[2],
+        date[3],
+        fileName
+      )
+      stream.emit('upload', path, s3.bucket + '/' + key)
+      s3.putObject({
+        Bucket: s3.bucket,
+        Key: key,
+        Body: fs.createReadStream(path)
+      })
+        .on('httpUploadProgress', (progress) => {
+          stream.emit(
+            'uploading',
+            path,
+            s3.bucket + '/' + key,
+            progress)
+        })
+        .on('success', () => {
+          stream.emit('uploaded', path, s3.bucket + '/' + key)
+          // remove the uploaded file
+          return _unlink(path, stream)
+            .then(() => {
+              resolve('S3')
             })
-        }
-
-        // all is well, write the data
-        fs.WriteStream.prototype._write
-            .call( that, data, encoding, function ( err ) {
-                cb( err )
-            });
-    })
-}
-
-RotateStream.prototype._compress = function ( path ) {
-    var that = this;
-    if ( path ) {
-        this._compressq.push( path );
-        if ( this._compressq.length > 1 ) {
-            return; // compression already running
-        }
+        })
+        .on('error', () => {
+          stream.emit('error')
```

Again - this will cause two error events, one here and the other in `_write`. It's a point I already described previously.  Also - you're swallowing the actual error. This will be difficult to debug because you're not transmitting any information about the original error, instead you create a new error and emit an empty error event.

---

Removed duplicate `emit('error')` calles